llvm/lib/IR/Type.cpp

```diff
@@ -210,8 +216,13 @@ int Type::getFPMantissaWidth() const {
   if (getTypeID() == DoubleTyID) return 53;
   if (getTypeID() == X86_FP80TyID) return 64;
   if (getTypeID() == FP128TyID) return 113;
-  assert(getTypeID() == PPC_FP128TyID && "unknown fp type");
-  return -1;
+  // See comment in the declaration of Type::getFPMantissaWidth();
+  // it returns -1 if the FP type does not have a stable mantissa.
```

Comment:
I think it's worth adding a clarification in Type.h documentation that `getFPMantissaWidth` returns -1 for decimal floating point types.

From looking at the existing calls to this method, it's safer to return -1 here than the number of decimal digits that decimal FP types can store. The main uses of this function are a) what's the largest integer that can be precisely stored in this FP types and b) checks for when double rounding is known to not be an issue, and in both cases, bailing out for decimal FP types is the correct thing to do.

Reply:
Should this function support DFP types at all? Perhaps we should just `assert(!isDecimalFloatingPointTy())`.

Reply:
Are we not going to need to compute the mantissa for a DFP, you think?

Reply:
I think we will, but whether that number makes sense in the context of the existing callers is unclear to me. I would expect existing callers to interpret the result in a context that assumes binary FP and therefore draw incorrect inclusions about precision. Perhaps this is a function that it would make sense to split for binary FP and decimal FP types.

Reply:
Ok. Creating a getDFPMantissaWidth.

Reply:
Sounds good. Should we rename the existing function to `getBFPMantissaWidth()` to make it explicit that it is only for binary FP?

Reply:
I'd shy away from renaming for now; this is a topic that is best visited during the ultimate community code review rather than among ourselves right now.